### ISSUE TITLE
Aeotec SmartSwitch 7 EU

### ISSUE THREAD

**Issue by Olsen93** [jvmahon/HubitatCustom#2]
Using the Zwave Plus Switch drivers version from 14.12.2020 Aeon SS7 will not be found in the database.
Manufacturer ID: 0371
Label: ZW175
![ss7](https://user-images.githubusercontent.com/26522853/102205280-47463a00-3ecb-11eb-838e-bd4c4e3fca3b.png)

**Comment by jvmahon** [comment]
I'll work on some updates so that, at the very least, I catch the error more gracefully. 

This is a relatively new device, so it may be a few weeks until the opensmarthouse.org database is updated.

**Comment by Olsen93** [comment]
Thanks! I keep watching the forum for news and will report back then.

**Comment by jvmahon** [comment]
Try again using the updated drivers at https://github.com/jvmahon/HubitatCustom

The opensmarthouse.org database looks like it does have an entry for this device. https://www.opensmarthouse.org/zwavedatabase/1134

**Comment by Olsen93** [comment]
Thank you!
Unfortunately it doesn't work, the logs seem to look different.
![ss7_2](https://user-images.githubusercontent.com/26522853/102747824-46952400-4361-11eb-8f11-2a8043d6fe77.png)
Here's the device data:
![data](https://user-images.githubusercontent.com/26522853/102747976-97a51800-4361-11eb-9ef8-55aaa3036bd8.png)

**Comment by jvmahon** [comment]
Try updating to the latest driver on the site, then click the button “Reseet Driver State Data” then “Initialize”
Do the logs look OK? - Particularly watching for the “Error” message?

Then, either way, do a restart of Hubitat (if you use the driver for multiple devices, this should reset the others).

**Comment by Olsen93** [comment]
Still won't work. Here's what I did:
- replaced the driver
- rebooted the hub
- clicked “Reset Driver State Data” then “Initialize”

![ss7_3](https://user-images.githubusercontent.com/26522853/102765429-e4e2b300-437c-11eb-97d1-743462ff7789.png)

**Comment by jvmahon** [comment]
Ugh. Its always the firmware process that fails! Sometimes devices just don't respond in time.

Is there anything at all in your "state" data on the device web page?

Select "initialize" a few times (without resetting the state data). If it succeeds once, then you are set as it should then store the firmware in the "state" data.

I'm planning a minor re-write of the firmware function. In the meantime,  I can put in a "fix" for you which can force a specific firmware version that should then solve this problem. Probably take a day.

**Comment by Olsen93** [comment]
> Is there anything at all in your "state" data on the device web page?

Yes, there are values for energy, power, switch and voltage. Not sure if these data had been retrieved from the switch or it is historical data from the former driver. 
Did I get this right, that these data is supposed to be deleted when clicking  “Reset Driver State Data”?
I can tell that it won't.


> Select "initialize" a few times (without resetting the state data)

Tried this 4 times in a row after another reboot, but it didn't help unfortunately.
Maybe you can give me some timing info, like how long it may take to reset or initialize, just to make sure that I am simply not too impatient.



> In the meantime, I can put in a "fix" for you which can force a specific firmware version

As for me, you can take a break if you want, but if you want me me to test, I'm your man, I really appreciate your efforts!

**Comment by jvmahon** [comment]
I may have found the error - the code didn't properly handle securely paired devices and didn't configure properly when the device data includes secureInClusters. I think (hope) I fixed that problem.
Update to the version that I posted this evening (Dec. 22) and try again.

**Comment by Olsen93** [comment]
You nailed it! Querying the device took about 5 seconds.
Identified issues so far:
1. during initializing the follwing messages show up:
 `error java.lang.NullPointerException: Cannot get property 'kWh' on null object on line 962 (initialize)`

   `warn Received a V2 Basic Report with mismatched value and targetValue and non-zero duration: BasicReport(value:255, 
     targetValue:0, duration:0)
`
2. when clicking "Refresh" a basic set command with value 255 seems to be sent, which toggles the switch off. In the device events a "physical off" will be reported. This also happens after rebooting HE or saving preferences.
3. under "Current States" the Ampere value won't show up, it is being reported in the device events, though.
4. "State Variables" only show parameterData. These seem to be correct.
5. The buttons to change parameters in the GUI under "Preferences" don't show up.
6. Clicking "Refresh" won't refresh the sensor values. This worked with the previous version, at least I found some event entries.
7. Sensor's won't report any values in configured time interval.
8. Trying to change a parameter results in this message: 
```
 error groovy.lang.MissingMethodException: No signature of method: user_driver_jvm__Beta__Advanced_Zwave_Plus_Metering_Switch_1122.setParameter() is applicable for argument types: (java.math.BigDecimal, java.math.BigDecimal, java.math.BigDecimal) values: [111, 4, 1800]
Possible solutions: setParameter(), setParameter(java.lang.Short), setParameter(java.lang.Short, java.lang.Short), setParameter(java.lang.Short, java.lang.Short, java.math.BigInteger) (setParameter)
```
Far from being a developer this is the first time for me working with Github, so if you want bug reports in a different format, please let me know.
But thanks for your great work so far!

**Comment by jvmahon** [comment]
All very helpful comments.  Most should be pretty straightforward to address.
As for sensors, that code is definitely not ready - its just a skeleton of functions.
Also, parameter settings are broken.
Probably will take a few days.

**Comment by Olsen93** [comment]
Thank you! 
Just replaced the driver, rebooted, and found that everything looked promising from the start, enabled debug mode and initialized again. No errors or warnings at all. 

1. fixed it.
2. this also didn't happen again, neither when refreshing, nor by saving or rebooting. But since I don't have physical access to the device atm, I can't tell if it has been really turned off. 
4. State Variables seem complete now
5. fixed it. It has been turned on right after installing the new driver, don't know how it was before, but I never touched that slider. Buttons are shown now.
7. found entries for  V/A/KWh in the eventlog from 19.12. - 20.12.2020, so it may have worked with this version, but I'm not sure if I reverted to the Aeon SS6 drivers at these days, so handle this information with care. 

May I suggest to do a minor version counting up, I think this could simplify communication over time?
Also, is there a way to prevent debug logging from turning off after a certain time?

**Comment by jvmahon** [comment]
Thanks.
Yes, I plan to fix the version # system. Right now it's really just used as a trigger to clearing the state. Any change in version # (whether an increase, decrease, etc.) erases the saved state and causes a re-initialization. I plan to put in some type of system where only "major" number of changes cause that to happen, while minor # changes don't.

**Comment by Olsen93** [comment]
Spotted and installed the new driver, do you already want to hear about my findings?

**Comment by jvmahon** [comment]
Sure, I am already working on an update that I'll release in the next few days, so its a good time to let me know of issues.
As an FYI - I plan to release the next version on Hubitat Package Manager.

**Comment by Olsen93** [comment]
> I plan to release the next version on Hubitat Package Manager.

This is great news, thanks! 
When doing a driver match-up using "Package Manager" a driver called "Advanced Universal Z-Wave Drivers" will be found.
I don't think it's the right one?

Reporting for driver version 0.0.2, referring to my original list of found issues:
3. amperage now shows up under "Current States", but disappers when reloading the page. Clicking "Refresh" or "Meter Refresh" makes it show up again. Using Chrome.
6. Fixed, clicking "Refresh" or "Meter Refresh" will now update the meter's values.
7. Meters are still not reporting, neither in the configured time interval nor according to the configures threshold.
8. Fixed. Parameters can be set now. Comparing with the "Basic Z-Wave Tool" driver shows that changes are being written to the device.
9.  New finding: In "State Variables" there are two persisting pending changes for parameter 255 (Factory reset) and 19 (Start/Stop LED flash). Checking them with the "Basic Z-Wave Tool" reports the value for parameter 4 for each of them.

**Comment by jvmahon** [comment]
Yes, that's the right package, but I haven't yet checked to make sure package manager handles the drivers properly.

Meters are still a work-in-progress. I'm working on them over the next few days.

**Comment by Olsen93** [comment]
So I will keep looking out for new driver versions here until this is official.
Did you have a look at the forums, there's another guy having trouble  with setting parameter 19, in a different way with a differnet device, though?

**Comment by jvmahon** [comment]
Yes, I noticed that. I'll be looking into that problem too.

**Comment by Olsen93** [comment]
10. Userguide is placed in the middle of State Variables. Have already seen this with v0.0.1, but not in 0.0.2
![manual](https://user-images.githubusercontent.com/26522853/103403676-4eb74580-4b51-11eb-9cd0-2d22e1e0f8c3.png)

**Comment by jvmahon** [comment]
That was intentional, but its over-crowded, so I plan to remove the inclusion / exclusion info. and, instead, put a link to the device's web page on opensmartesthouse.org

**Comment by jvmahon** [comment]
0.0.5 posted - this should fix some of the problems setting parameters.

**Comment by Olsen93** [comment]
> 9. New finding: In "State Variables" there are two persisting pending changes for parameter 255 (Factory reset) and 19 (Start/Stop LED flash)

This issue still persists with v0.0.5.
Did a reboot after replacing the driver, reseted driver state data and re-initialized.

**Comment by jvmahon** [comment]
Try 0.0.6

**Comment by Olsen93** [comment]
Reporting for 0.0.7.
3. amparage now has completely gone in"Current States". This seems to be the related error message: "Received unexpected meter type for Kristall. Only type '1' (Electric Meter) is supported. Received type: 33".
6. Meters are not reporting at all again, even not when clicking "(Meter) Refresh". 
9. Pending changes were gone right after the reboot, but returnend after saving preferences.
Additionally, new parameter 82 (Night Light Mode) showed up, with the value 301991936.
Recalculating it I found this is a valid value for turning on at 06pm and off at 08am.
Never noticed that before, also I did a factory reset of the switch before including and never used the night light with it.

**Comment by Olsen93** [comment]
> Additionally, new parameter 82 (Night Light Mode) showed up, with the value 301991936.
Think I have to apologize for this, seems to be a HE problem. Checking this parameter with the "Basic Z-Wave Tool" I found out that this applies to all of my 4 Smart Switches, parameter 82 cannot be changed to "0" to any of them, still the night light function doesn't seem to be activated.

**Comment by jvmahon** [comment]
No worries. The feedback is appreciated. I did make a lot of changes to the meter handling code so it can work with devices having endpoints, so there may have been some errors as I did that.  

Meters may not work right until after 2.2.5 of Hubitat - I've encountered a few errors in the zwave parser which are causing problems (the Hubitat staff has already identified fixes that are needed). I suspect the self-configuring features I've included are pushing the boundaries of what most people have included in their code, so I'm encountering bugs in Hubitat that may not have been noticed before.
